lua/lvim/plugin-loader.lua

```diff
@@ -3,174 +3,178 @@ local plugin_loader = {}
-    compile_path = compile_path,
-    snapshot_path = snapshot_path,
-    max_jobs = max_jobs,
-    log = { level = "warn" },
-    git = {
-      clone_timeout = 120,
-    },
-    display = {
-      open_fn = function()
-        return require("packer.util").float { border = "rounded" }
-      end,
-    },
-  }
-
-  if in_headless then
-    init_opts.display = nil
-    init_opts.git.clone_timeout = 300
-  end
-
-  if not utils.is_directory(install_path) then
+  if not utils.is_directory(lazy_install_dir) then
     print "Initializing first time setup"
-    print "Installing packer"
-    print(vim.fn.system { "git", "clone", "--depth", "1", "https://github.com/wbthomason/packer.nvim", install_path })
-    vim.cmd "packadd packer.nvim"
+    vim.fn.system {
+      "git",
+      "clone",
+      "--filter=blob:none",
+      "--single-branch",
+      "https://github.com/folke/lazy.nvim.git",
+      lazy_install_dir,
+    }
   end
 
-  local status_ok, packer = pcall(require, "packer")
-  if status_ok then
-    packer.on_complete = vim.schedule_wrap(function()
-      require("lvim.utils.hooks").run_on_packer_complete()
-    end)
-    packer.init(init_opts)
-  end
-end
+  remove_rtp_paths()
 
--- packer expects a space separated list
-local function pcall_packer_command(cmd, kwargs)
-  local status_ok, msg = pcall(function()
-    require("packer")[cmd](unpack(kwargs or {}))
+  vim.opt.runtimepath:prepend(lazy_install_dir)
+
+  -- Add plugins to rtp (needed for config:init)
+  -- TODO: is there a better way to do this?
+  require("lazy.core.util").ls(plugins_dir, function(path, name, type)
+    if type == "directory" and name ~= "lazy.nvim" then
+      vim.opt.rtp:append(path)
+    end
   end)
-  if not status_ok then
-    Log:warn(cmd .. " failed with: " .. vim.inspect(msg))
-    Log:trace(vim.inspect(vim.fn.eval "v:errmsg"))
-  end
 end
 
-function plugin_loader.cache_clear()
-  if not utils.is_file(compile_path) then
-    return
-  end
-  if vim.fn.delete(compile_path) == 0 then
-    Log:debug "deleted packer_compiled.lua"
-  end
-end
+function plugin_loader.reload(spec)
+  local Config = require "lazy.core.config"
+  local lazy = require "lazy"
 
-function plugin_loader.compile()
-  Log:debug "calling packer.compile()"
-  vim.api.nvim_create_autocmd("User", {
-    pattern = "PackerCompileDone",
-    once = true,
-    callback = function()
-      if utils.is_file(compile_path) then
-        Log:debug "finished compiling packer_compiled.lua"
-      end
-    end,
-  })
-  pcall_packer_command "compile"
-end
+  -- TODO: reset cache?
 
-function plugin_loader.recompile()
-  plugin_loader.cache_clear()
-  plugin_loader.compile()
-end
+  Config.spec = spec
 
-function plugin_loader.reload(configurations)
-  _G.packer_plugins = _G.packer_plugins or {}
-  for k, v in pairs(_G.packer_plugins) do
-    if k ~= "packer.nvim" then
-      _G.packer_plugins[v] = nil
-    end
+  require("lazy.core.plugin").load(true)
```

Comment:
you really shoudn't do all of this.

lazy automatically installs missing plugins during startup

Reply:
That's for hot reload, so it will install missing plugins when the config is saved (we have an autocommand)

Reply:
ah ok. Lazy also does change detection that works even when you make changes outside of Neovim. I can provide a user event that triggers when changes are detected.

Reply:
that automatic change detection requires polling a lua module as far as I know, and currently lunarvim and user plugins are split, they could be joined, but the polling would be a bit expensive, since I currently parse a json file to set `commit` in plugin specs, we are working on packaging lunarvim and that step could be done during packaging and then it wouldn't be a problem, but we need the autocommand anyway and doing it here also means no delay, if we switch to a lua module for the specs, we could disable change detection timer and trigger checking from the autocmd, I'm not sure if it's worth the effort though